### PR TITLE
copy source into prefix

### LINKED ISSUE #4083: Add global 'source' variant to install source code
Sometimes it would be useful to have the source code together with the build, so I would like to add a variant (`+source`) in somewhere in the package base class (`spack/spack/package.py`) to trigger a copy of the source code (after patching, before configure/cmake/build) to `<prefix>/src`.

Any suggestions?

### PR DESCRIPTION
**Pull request by junghans**
Fix #4083.

### CODE REVIEW

File: lib/spack/spack/package.py [hunk 1]
```diff
@@ -1165,6 +1166,8 @@ def do_install(self,
             keep_stage (bool): By default, stage is destroyed only if there
                 are no exceptions during build. Set to True to keep the stage
                 even with exceptions.
+            install_source (bool): By default, source is not installed, but
+                for debugging it might useful to keep it around.
```
Comment (davydden) [inline lib/spack/spack/package.py]: Might be useful

File: lib/spack/spack/package.py [hunk 2]
```diff
@@ -1263,6 +1267,10 @@ def build_process(input_stream):
 
             self.stage.keep = keep_stage
             with self._stage_and_write_lock():
+                if install_source:
+                    src_target = join_path(self.spec.prefix, 'src')
+                    tty.msg('Copying source to {0}'.format(src_target))
+                    install_tree(self.stage.source_path, src_target)
```
Comment (adamjstewart) [inline lib/spack/spack/package.py]: You'll probably want to check if the src directory actually exists first.

Comment (junghans) [inline lib/spack/spack/package.py]: Done

**Comment by junghans** [comment]
> 2. Can you make the necessary modifications for this to be runnable after an existing install? 

How would I do that?

**Comment by junghans** [comment]
@citibeth: see https://github.com/LLNL/spack/issues/4083#issuecomment-298929264

**Comment by junghans** [comment]
@tgamblin: I have done 1.) and 4.), I like having `src` inside the prefix as source could be patched different for different variants, so it is bit more clear.

However for 2.) I will need some pointer in the code. ( @adamjstewart / @tgamblin )

File: lib/spack/spack/test/install.py [hunk 3]
```diff
@@ -58,6 +58,25 @@ def fake_fetchify(url, pkg):
 
 
 @pytest.mark.usefixtures('install_mockery')
+def test_source_copy(mock_archive):
```
Comment (junghans) [inline lib/spack/spack/test/install.py]: @adamjstewart : Is this test run?

Comment (adamjstewart) [inline lib/spack/spack/test/install.py]: It should run. All functions that start with `test_` should be run. I'm not sure how often Codecov updates, if that's what you're asking.

Comment (junghans) [inline lib/spack/spack/test/install.py]: Hmm and the `source_install=True` below should trigger the new code path or does the `mock_archive` doesn't have source?

**Comment by junghans** [comment]
Rebased: ping @tgamblin

**Comment by junghans** [comment]
Ping @tgamblin

**Comment by junghans** [comment]
Ping @tgamblin

**Comment by tgamblin** [comment]
fixed flake8 errors.

**Comment by tgamblin** [comment]
Added a test.  @junghans: it's easier to do now than it was when you added this :)

**Pull request merged**